src/server/engine_shard_set.cc

```diff
@@ -80,20 +80,22 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
 // 3. in case the above is OK, make sure that we have a "gap" between usage and commited memory
 // (control by commit_use_threshold flag)
 bool EngineShard::DefragTaskState::IsRequired() const {
-  const uint64_t threshold_mem = max_memory_limit * GetFlag(FLAGS_mem_defrag_threshold);
+  const int64_t threshold_mem = max_memory_limit * GetFlag(FLAGS_mem_defrag_threshold);
   const double commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
 
   if (cursor > kCursorDoneState) {
     return true;
   }
 
-  uint64_t commited = GetMallocCurrentCommitted();
+  // can be negative due to weird accounting of mimalloc.
+  int64_t commited = GetMallocCurrentCommitted();
+
   uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
 
   // we want to make sure that we are not running this to many times - i.e.
   // if there was no change to the memory, don't run this
   if (threshold_mem < commited && mem_in_use > 0 &&
-      (uint64_t(mem_in_use * commit_use_threshold) < commited)) {
+      (uint64_t(mem_in_use * commit_use_threshold) < uint64_t(commited))) {
```

Comment:
If you comment that `commited` can be negative... then why do you raw cast everything to uint64_t? It'll underflow and become a very large value

Reply:
i check before that `threshold_mem < commited`  and since threshold_mem is always positiive, then i verify that commited is also positive.